### ISSUE TITLE
[Proposal] API for state store filtering and querying

### ISSUE THREAD

**Issue by vaclavpetr** [dapr/dapr#1339]
## In what area(s)?

/area runtime

## Describe the feature
We would like to use state store as main storage through Dapr.Client in SDK (for us .NET SDK), but state store APIs allow only single GET request for one stored document.
It would be great, to have opportunity to retrieve more values by querying, like OData described in Cosmos DB in [MSDN documentation](https://docs.microsoft.com/cs-cz/azure/cosmos-db/tutorial-query-table) (or another REST queriyng language):
https://<mytableapi-endpoint>/People()?$filter=PartitionKey%20eq%20'Smith'%20and%20Email%20eq%20'Ben@contoso.com'

Example of DAPR API call:
GET http://localhost:3500/v1.0/state/people-storage?$filter=PartitionKey%20eq%20'Smith'%20and%20Email%20eq%20'Ben@contoso.com'

It would be great to retrieve only keys by some condition.
Example of DAPR API call:
GET http://localhost:3500/v1.0/state/people-storage?getAll=keys

## Usage
For daily basis, state management have missing querying for usage as main storage. For our usage, we need use custom SDK in our project, even if we are using Dapr runtime, because it's possible to make more sophisticated filtering calls for getting more than one record, and filter by some conditions.

In the current situation, it's possible to only to call more GetAsync in parallel for retrieving all values.

**Comment by benc-uk** [comment]
This is sorely needed, I don't see how the state service in Dapr is of any real world use with out some means to query/filter

**Comment by jjcollinge** [comment]
+1 I’d like to see OData support for the state API. It’d make dapr far more applicable to users. cc: @fbeltrao what are your thoughts on this?

**Comment by fbeltrao** [comment]
I agree that adding state would enable more scenarios. I am not sure how many features we would have to add in order to create a state component that offers basic database features.

For instance, to properly query/update based on specific columns you might have to add indexes. Would that be part of the component definition? Where do we stop?

Maybe a way to look at it is find out which scenarios we should enable. I am not sure all state related scenarios should be solved by Dapr. 

Perhaps, building sample application where a component connects directly to a database also breaks this idea that in order to use Dapr you should only use Dapr.

**Comment by jjcollinge** [comment]
@fbeltrao yeh the Dapr API definitely shouldn't try to be all things to everybody and should minimize its footprint where it can and it should look to leverage existing standards, such as OData.

That being said, if Dapr's State API doesn't support querying and aggregation, people **will** work around the constraint using eventing and aggregation/query services which in _most_ cases are an unnecessary complexity or they'll just drop Dapr completely.

I think if you look at the subset of [OData that Azure Table Storage](https://docs.microsoft.com/en-us/rest/api/storageservices/querying-tables-and-entities#supported-query-options) supports you get a good idea of what might be needed i.e.

- $filter
- $top
- $select

With the following operators:

- Equal
- GreaterThan
- GreaterThanOrEqual
- LessThan
- LessThanOrEqual
- NotEqual
- And
- Not
- Or

**Comment by fbeltrao** [comment]
@jjcollinge having these features would be great. It solves the problem of getting a data collection. Then comes my follow up questions:
- How would database index be defined? Should we add it as part of the component description? Depending on the underlying database, it might requires additional optimizations to return results with an acceptable performance. 

Moveover, I have seen folks asking for a way to update multiple documents at once using Dapr (i.e. `UPDATE users SET x=1 WHERE y=2`). Would that be out of scope for now?

**Comment by jjcollinge** [comment]
@fbeltrao I think considering indexes is probably worth it's own discussion outside the scope of OData - However, updating a collection is [defined by the OData spec](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_UpdateMembersofaCollection) but may be something to look at further down the line. From my initial searching, it doesn't look like there's a good candidate 3rd party OData v4 library written in Go. Which means we'd likely have to write all the query parsing logic in Dapr. Given this, it'd best to start small and like you said, target specific use cases 1 at a time. I'd consider the basic querying (filter, top, select) to be the highest priority.

**Comment by jjcollinge** [comment]
So to solidify this a little more, we would need to add odata parsing logic to parse the querystring and construct an AST. I played with [godata](https://github.com/crestonbunch/godata) which appears to have been abondoned but atleast lays some ground work, it could parse `$filter`, `$top` and `$select` fine. We'd then need a set of query builders that could take the AST and build a query in a given dialect. We could parse the querystring in `onGetState` and pass the query AST as an optional property on the `state.GetRequest` object. The provider can then use the appropriate query builder to convert the AST into a query that it can then submit to it's state provider. If we want to keep all the odata code together, it feels like this should probably live under `components-contrib/state/odata`. If a state provider doesn't support OData and a reuqest tries to use it then it return a bad request with error message.

**Comment by msfussell** [comment]
@jjcollinge Can you propose a design for this and then we can take this from there?

**Comment by jjcollinge** [comment]
@msfussell yeh I can look into it.

**Comment by jjcollinge** [comment]
### TLDR:
I'm not yet convinced this is a good idea!

> NOTE: This is WIP!
# Why?
Dapr's current state API facilitates the retrieval of an item from a state store only by a given key. Storing and retrieving values based on a key is essentially the purpose of a key-value databases. This makes them very useful for a certain type of isolated and highly parallel data access pattern, such as a shopping cart, session state or caching. As the capability to store data indexed by a key is fundamental to storage providers, it is technically possible to use many different storage technologies as a key-value store, although some are highly tuned for this type of data storage.

In my experience, _most_ problems require more advanced state capabilities than simple key-value storage. Therefore, the dapr state API wouldn’t be applicable to _most_ state management problems. That alone is not a reason to justify adding more complex behaviour to the dapr state API. However, if _most_ users must use an existing database provider’s SDK, then _most_ users would be ok to use said SDK for key-value storage as well rather than using dapr’s state API too. I’d say It’s fairly unlikely that a lot of users have requirements that would drive them to use a specific key-value data technology via dapr and another database SDK too.

I would also suggest that constraining complexity from the API merely moves it to the user’s architecture. I’ve seen a number of exotic architectures used to work around such constraints. Such as sending ‘n’ parallel requests to fetch all the data and then filtering locally, or having many services maintaining secondary indexes via eventing or even polling. We want dapr to encourage simple architectural design rather than forcing users to “get creative”.

Dapr now has support for many different state providers that are not primarily key-value stores. Furthermore, it is the minority of state providers who strictly do not provide more advanced features.
It’s worth nothing that different data technologies have different data models and therefore have different capabilities. For instance, key-value stores typically provider filtering on only the primary key, whereas document stores query the entire document.

It should also be noted that performing scans on key-value stores is usually O(N) unless using secondary indexes. This should be taken into account when considering the capabilities of each state provider.

**Aerospike**
Aerospike is a key-value store with supports for scans, querying (AQL), secondary indexes and UDFs. Therefore, Aerospike can perform select, where, top, count, range, geospatial, filters and aggregation.
References:
-	https://www.aerospike.com/docs/guide/scan.html
-	https://www.aerospike.com/docs/tools/aql/index.html
-	https://www.aerospike.com/docs/client/go/usage/query/query.html
-	https://www.aerospike.com/docs/guide/query.html

**DynamoDB**
DynamoDB supports finding items by primary key using key conditions, filter conditions, limits and count.
References:
- https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Query.html#Query.FilterExpression
- https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Query.html 

**Azure CosmosDB**
Azure CosmosDB supports multiple data models, but we use it as a document store which can be queried using a dialect of SQL.
References:
-	https://github.com/a8m/documentdb#querydatabases

**Azure Table Storage**
Azure Table Storage is a hierarchical key-value store which can be queried at the table, entity or entity property level. These queries can be express as a subset of OData.
References:
-	https://docs.microsoft.com/en-us/rest/api/storageservices/querying-tables-and-entities

**Cassandra**
Cassandra supports a query language called CQL that models data as tables with rows and columns. CQL supports  select, where, count and limit semantics. CQL can be used to execute distributed queries and also has support for working with JSON documents. 
References: http://cassandra.apache.org/doc/latest/cql/

**Cloudstate**
As far as I can see you cannot perform queries across Cloudstate.

**etcd**
Etcd is a key-value store. It’s API supports range-based filtering, count and limit on keys.
References:
-	 https://github.com/etcd-io/etcd/blob/master/Documentation/learning/api.md

**CouchDB**
CouchDB is a document database and supports a JSON query syntax that supports selectors, limit, skip, and sort.  
References:
-	 https://docs.couchdb.org/en/stable/api/database/find.html

**Consul**
Consul has a hierarchical key-value store which allows you to retrieve collections of keys using paths.
References:
-	 https://www.consul.io/api/query.html

**HazelCast**
HazelCast supports distributed queries which can use a subset of a SQL and can query JSON documents.
References:
-	https://docs.hazelcast.org/docs/latest-dev/manual/html-single/#distributed-query

**Memcached**
No, Memcached doesn’t support any querying.

**Mongodb**
MongoDB is a document database and supports querying across documents using it’s own query syntax.
References:
-	 https://docs.mongodb.com/manual/tutorial/query-documents/

**Redis**
Redis can support filtering using keys, using secondary indexes and set operations. There is also a SQL like experience built of redis using RediSQL.
References:
-	https://redis.io/topics/indexes
-	https://github.com/RedBeardLab/rediSQL
-	https://redis.io/commands/keys

**SQLServer**
SQLServer supports a full relational querying capability expressed in various SQL dialects.

**Zookeeper**
No, Zookeeper doesn’t appear to support querying across data.

So we can roughly split the above state providers into 3 categories; key-value, document, relational.

# Why OData?
Dapr exposes a HTTP REST API and a gRPC API. This proposal works within those constraints, therefore, additional APIs such as GraphQL have been ruled out.
OData is a standard specification for defining a query as part of a HTTP request URI for REST APIs. There are well documented pros and cons to OData and I’m sure everyone has an opinion on it. However, the alternative is writing a custom query specification and implementation which seems like a fool’s errand.
Defining a query specification that would make sense across both the HTTP API and gRPC API would require too many levels of abstraction. Therefore, I think the OData specification can provide a enough query syntax that we can use in both APIs.

# How?
This change would mean that you could add an OData query to the HTTP or gRPC request. For HTTP this would result in the following URIs:
`http://localhost:<daprPor>/v1.0/state/<storename>/<key>?<odata-querystring>`

The following flow describes how this query can be handled by dapr and the state providers.

__HTTP:__
1.	The client sends a HTTP request with an OData query URL encoded as the querystring.
2.	The HTTP GetState API extracts the OData query from the HTTP requests querystring.
3.	The HTTP GetState API parses the OData query into an abstract syntax tree (AST) representation of the OData query.
4.	If the HTTP GetState API fails to parse the OData query, it returns an error.
5.	The HTTP GetState API passes the AST to the state store provider.
6.	The state provider can ignore the AST or error if it doesn’t support querying.
7.	If the state provider does support querying it has to implement a query builder which can convert the OData AST into a query specifically for that provider i.e. SQL.
8.	If the query builder fails to translate the OData AST into a native query, it returns an error.
9.	The state provider then executes the native query against the database and returns the results.

__gRPC:__
1.	The gRPC client sets a string property on the GetStateRequest to an OData query.
2.	The gRPC GetState API extracts the OData query from the gRPC GetStateRequest object.
3.	Steps 3 – 9 are the same as the HTTP flow.

This would mean that the support for querying would be _different_ per state provider and would introduce inconsistencies to the API. Switch state provider may now result in errors.

## Code changes required
\+ `components-contrib/state/odata/queryoptions.go` (parses a OData query string into an OData AST)
\+ `components-contrib/state/odata/queryoptions_test.go` (tests the above)
~ `dapr/pkg/grpc/api.go` (add OData query handling)
~ `dapr/pkg/http/api.go` (add OData query handling)
Any state provider that wants to use the query would then need to implement add the following:
\+ `components-contrib/state/<provider>/odata_querybuilder.go` (builds a <provider> query from an OData AST)
~ `components-contrib/state/<provider>/<provider.go>` (Calls the query builder and executes the query)

## OData Parsing and query building
The following OData query:
`filter=CountryCode eq 'US' and FirstName eq 'Alice'`

Would result in the following AST:

![image](https://user-images.githubusercontent.com/7241190/79361622-74f27200-7f3d-11ea-9f6c-ca552ba5f9fc.png)

The state provider would have to implement a query builder to traverse this tree and build up a query in their native syntax i.e.

`SELECT * WHERE CountryCode=’US’ AND FirstName=’Alice’`

However, if this AST was provided to a key-value provider, we assume that we’re only filtering on the keys and therefore this wouldn’t make sense.

This raises a few important questions. Firstly, should dapr differentiate between key-value state providers and other types i.e. document and relational. Secondly, would constraining dapr to _only_ filtering on keys (even for document and relational stores) give users a sufficient experience.

## Recommendation
This topic needs a much broader conversation about the goals of Dapr before I’d be confident this is a good idea. Currently, I have pretty strong concerns that having an inconsistent API across the different state providers is a bad idea. I also have reservations about introducing unnecessary complexity to the dapr API when users could easily use the native SDKs to get specialized behaviour.

To be continued…

**Comment by yaron2** [comment]
Thanks @jjcollinge for this really detailed proposal, context and explanation.

/cc @msfussell

**Comment by rdiaz82** [comment]
Hi!! thanks, @yaron2 for pointing me to this proposal.  A couple of weeks ago, I wrote another proposal very similar to this one but using bindings and a mongo based query language here is the proposal for more information https://github.com/dapr/components-contrib/issues/441. 

After reading the awesome detailed explanation done by @jjcollinge, I think that we can blend both proposals to solve the query problem for the state store. 

I see three different approaches that can be explored:

1. Use binding for the implementation and OData as a query language.  In this case, inside the `Invoke` method we can perform the parse of the OData string and just convert it to each particular database query language. In this approach, we can avoid the intermediate parse to get an AST query. This solution is simply replicate the current binding implementation for each database engine

2. Use binding and develop a parser from the OData string to get the AST which is passed to a method that transforms the AST into the particular DB query language and executes it. In this case, we can develop a specific package the current binding interface, and the Odata parsing to AST, just leaving the AST parsing to the specific engine to each database binding component.

3. Create a new database component based on bindings. In a very similar way that is proposed in `2`,  implement the parse from OData string to AST and as @jjcollinge proposed pass the AST to each particular database package where the AST is parsed and the query is executed.

In proposal `1` and `2`, the core of Dapr doesn't need to be modified because all the work is performed on the binding side. In the case of the proposal `3`, we will need to modify each database packages to add the method to parse from AST to the specific DB engine and perform the query. 

@jjcollinge @yaron2 what do you think about this?

**Comment by yaron2** [comment]
Hey @rdiaz82, I prefer option 1 for simplicity and maintainability.

Option 3 carries too much overhead.

What are the benefits of option 2 over option 1 in your eyes?

**Comment by artursouza** [comment]
This feature would require some sort of mapping between the OData syntax and the underneath querying language for the database. Depending on the query, it might not be feasible to achieve full portability.

Another way to support this would be to have some basic querying done on the server side and full odata in memory for daprd. So, applications can use what makes more sense for their app.

More ideas on this are appreciated.

**Comment by lichen782** [comment]
> This looks good, but I am looking for same feature in another components, and If this is not for all storage components same, it's not possible to switch storage to storage. I understand, that for Cosmos DB or for MS SQL or Redis, it wont be working, right ?

true. This is trade-off. even SQLs have their dialects, which we back in 3 classic tier projects, were handled by ORM (jpa, mybatis, hibernate). So  I think essentially, dapr lacks an ORM, abstracting plain SQL to query objects.

**Comment by tanekim77** [comment]
Can't we have a special Relational Database support? Relational databases alrady have Entity Framework Core ORM that already support complex queries and updates. Rather than trying to support both non-relational and relational storages in single Dapr storage, I would like to see there are special type of storage support for relational database that utilize Entity Framework Core

**Comment by artursouza** [comment]
@tanekim77 Supporting a Relational Database as a native Dapr API is one of the possibilities but it does not exclude this issue. The idea here is to have basic query capabilities for state stores and not to have feature parity with SQL databases. This is an example of a solution in a noSQL database with similar inspirations: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/ql-reference.html

**Comment by AlseinX** [comment]
Disappointed to see the postponement of the most important feature that makes dapr minimum usable for most projects.

**Comment by artursouza** [comment]
> Disappointed to see the postponement of the most important feature that makes dapr minimum usable for most projects.

We are actively working on it. @dmitsh will soon share a proposal for this. I tagged to 1.6 since it is the target to get it done given the priorities we have in 1.5 and the time to collect feedback on the proposal. We understand this feature as a top priority for this semester.

Thanks.

**Comment by artursouza** [comment]
@AlseinX Please, take a look at the Query State API proposal in #3662

**Comment by msfussell** [comment]
Closing this API issue. Issues tracking Alpha->Stable for the API and for Query API implemented in each of the SDKs are tracked separately.
